src/EventListener/DataContainer/Theme.php

```diff
@@ -0,0 +1,379 @@
+use Symfony\Component\HttpFoundation\Session\SessionInterface;
+use Symfony\Component\Security\Core\Authentication\Token\Storage\TokenStorageInterface;
+
+class Theme
+{
+    /** @var RequestStack */
+    protected $requestStack;
+
+    /** @var SessionInterface */
+    protected $session;
+
+    /** @var BackendUser */
+    protected $user;
+
+    /** @var ImageFactoryInterface */
+    private $imageFactory;
+
+    /**
+     * Theme constructor.
+     *
+     * @param RequestStack          $requestStack
+     * @param SessionInterface      $session
+     * @param TokenStorageInterface $tokenStorage
+     * @param ImageFactoryInterface $imageFactory
+     */
+    public function __construct(
+        RequestStack $requestStack,
+        SessionInterface $session,
+        TokenStorageInterface $tokenStorage,
+        ImageFactoryInterface $imageFactory
+    ) {
+        $this->requestStack = $requestStack;
+        $this->session      = $session;
+        $this->user         = $tokenStorage->getToken()->getUser();
+        $this->imageFactory = $imageFactory;
+    }
+
+    /**
+     * Check permissions to edit the table
+     *
+     * @throws AccessDeniedException
+     */
+    public function onCheckPermission(): void
+    {
+        if ($this->user->isAdmin) {
+            return;
+        }
+
+        $request = $this->requestStack->getCurrentRequest();
+
+        // Check the theme import and export permissions (see #5835)
+        switch ($request->get('key')) {
+            case 'importTheme':
+                if (!$this->user->hasAccess('theme_import', 'themes')) {
+                    throw new AccessDeniedException('Not enough permissions to import themes.');
+                }
+                break;
+
+            case 'exportTheme':
+                if (!$this->user->hasAccess('theme_import', 'themes')) {
+                    throw new AccessDeniedException('Not enough permissions to export themes.');
+                }
+                break;
+        }
+    }
+
+
+    /**
+     * Add an image to each record
+     *
+     * @param array  $row
+     * @param string $label
+     *
+     * @return string
+     */
+    public function onAddPreviewImage(array $row, string $label): string
+    {
+        if (!$row['screenshot']) {
+            return $label;
+        }
+
+        $objFile = FilesModel::findByUuid($row['screenshot']);
+
+        if (null === $objFile) {
+            return $label;
+        }
+
+        $imageUrl = $this->imageFactory
+            ->create(TL_ROOT . '/' . $objFile->path, [75, 50, 'center_top'])
+            ->getUrl(TL_ROOT);
+
+        return Image::getHtml($imageUrl, '', 'class="theme_preview"') . ' ' . $label;
+    }
+
+
+    /**
+     * Check for modified style sheets and update them if necessary
+     */
+    public function onUpdateStyleSheet(): void
+    {
+        if ($this->session->get('style_sheet_update_all')) {
+            // todo: make updateStyleSheets() static
```

Comment:
Why static? In the best case, the Stylesheets handling would reside in its own services too 😄